lib/collection/src/collection/shard_transfer.rs

```diff
@@ -357,4 +373,95 @@ impl Collection {
             }
         }
     }
+
+    /// Whether we have reached the shard transfer soft limit based on the given incoming and
+    /// outgoing transfers.
+    pub(super) fn check_shard_transfer_limit(&self, (incoming, outgoing): (usize, usize)) -> bool {
+        self.shared_storage_config
+            .incoming_shard_transfers_limit
+            .map(|limit| incoming >= limit)
+            .unwrap_or(false)
+            || self
+                .shared_storage_config
+                .outgoing_shard_transfers_limit
+                .map(|limit| outgoing >= limit)
+                .unwrap_or(false)
+    }
+}
+
+/// A structure tracking ongoing and proposed shard transfers, usually used globally.
+///
+/// This can be used to count the number of shard transfers on some peer to rate-limit automatic
+/// transfers.
+#[derive(Debug, Default)]
+pub struct ShardTransferTracker {
+    /// Ongoing shard transfers.
+    transfers: Vec<ShardTransferKey>,
```

Comment:
This implementation currently relies on tracking the ongoing shard transfers ourselves.

I implemented the first iteration of this to be global, meaning the configured limits apply to the node across all collections. In that configuration tracking ongoing transfers here was required.

Since, I've changed this to be on collection level. Because of that, it isn't strictly required anymore to track transfers here. We could also rely on the list of transfers in the collection's shard holder.

I've not decided yet whether to replace this with ongoing transfers in the shard holder. It may be somewhat tricky to rework due to locking issues.

What do you think? Keep it as-is, or rework this to rely on our existing list of ongoing transfers?